File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/filter/patient/PatientByHcPartyGenderEducationProfession.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto.filter.patient;

import org.taktik.icure.entities.Patient;
import org.taktik.icure.entities.embed.Gender;
import org.taktik.icure.services.external.rest.handlers.JsonPolymorphismRoot;
import org.taktik.icure.services.external.rest.v1.dto.filter.Filter;

import java.util.Objects;

@JsonPolymorphismRoot(Filter.class)
public class PatientByHcPartyGenderEducationProfession extends Filter<Patient> implements org.taktik.icure.dto.filter.patient.PatientByHcPartyGenderEducationProfession {
    private String healthcarePartyId;
    private Gender gender;
    private String education;
    private String profession;

    public PatientByHcPartyGenderEducationProfession() {
    }

    public PatientByHcPartyGenderEducationProfession(Gender gender, String education, String profession, String healthcarePartyId) {
        this.healthcarePartyId = healthcarePartyId;
        this.gender = gender;
        this.education = education;
        this.profession = profession;
    }

    @Override
	public boolean matches(Patient item) {
        return (healthcarePartyId == null || item.getDelegations().keySet().contains(healthcarePartyId))
                && (gender == null || item.getGender() != null && item.getGender() == gender)
                && (education == null || item.getEducation() != null && item.getEducation().equals(education))
                && (profession == null || item.getProfession() != null && item.getProfession().equals(profession));
    }

    @Override
    public Gender getGender() {
        return this.gender;
    }

    @Override
    public String getEducation() {
        return this.education;
    }

    @Override
    public String getProfession() {
        return this.profession;
    }

    @Override
    public String getHealthcarePartyId() {
        return this.healthcarePartyId;
    }

    public void setHealthcarePartyId(String healthcarePartyId) {
        this.healthcarePartyId = healthcarePartyId;
    }

    public void setGender(Gender gender) {
        this.gender = gender;
    }

    public void setEducation(String education) {
        this.education = education;
    }

    public void setProfession(String profession) {
        this.profession = profession;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof PatientByHcPartyGenderEducationProfession)) return false;
        PatientByHcPartyGenderEducationProfession that = (PatientByHcPartyGenderEducationProfession) o;
        return Objects.equals(healthcarePartyId, that.healthcarePartyId) &&
                gender == that.gender &&
                Objects.equals(education, that.education) &&
                Objects.equals(profession, that.profession);
    }

    @Override
    public int hashCode() {
        return Objects.hash(healthcarePartyId, gender, education, profession);
    }
}

File: src/main/kotlin/org/taktik/icure/domain/filter/impl/patient/PatientByHcPartyGenderEducationProfession.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.domain.filter.impl.patient

import com.github.pozo.KotlinBuilder
import org.taktik.icure.domain.filter.AbstractFilter
import org.taktik.icure.entities.Patient
import org.taktik.icure.entities.embed.Gender
import java.util.*

@KotlinBuilder
data class PatientByHcPartyGenderEducationProfession(
        override val desc: String? = null,
        override val healthcarePartyId: String? = null,
        override val gender: Gender? = null,
        override val education: String? = null,
        override val profession: String? = null
) : AbstractFilter<Patient>, org.taktik.icure.domain.filter.patient.PatientByHcPartyGenderEducationProfession {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PatientByHcPartyGenderEducationProfession) return false
        return healthcarePartyId == other.healthcarePartyId && gender === other.gender &&
                education == other.education &&
                profession == other.profession
    }

    override fun hashCode(): Int {
        return Objects.hash(healthcarePartyId, gender, education, profession)
    }

    override fun matches(item: Patient): Boolean {
        return ((healthcarePartyId == null || item.delegations.keys.contains(healthcarePartyId!!))
                && (gender == null || item.gender != null && item.gender === gender)
                && (education == null || item.education != null && item.education == education)
                && (profession == null || item.profession != null && item.profession == profession))
    }
}
